From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 12emin34 <macanovic.emin@gmail.com>
Date: Sat, 30 Apr 2022 13:35:56 +0200
Subject: [PATCH] Zircon config files


diff --git a/src/main/java/me/_12emin34/zircon/ZirconConfig.java b/src/main/java/me/_12emin34/zircon/ZirconConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..f6bff62fcc97af95456efa8aae1d924b8b96deb6
--- /dev/null
+++ b/src/main/java/me/_12emin34/zircon/ZirconConfig.java
@@ -0,0 +1,146 @@
+package me._12emin34.zircon;
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import me._12emin34.zircon.command.ZirconCommand;
+import net.minecraft.server.MinecraftServer;
+import org.bukkit.Bukkit;
+import org.bukkit.command.Command;
+import org.bukkit.configuration.ConfigurationSection;
+import org.bukkit.configuration.InvalidConfigurationException;
+import org.bukkit.configuration.file.YamlConfiguration;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+
+@SuppressWarnings("unused")
+public class ZirconConfig {
+    private static final String HEADER = "This is the main configuration file for Zircon.\n"
+            + "Some options may impact gameplay, so use\n"
+            + "with caution, and make sure you know what each option does before configuring.\n";
+    public static YamlConfiguration config;
+    static int version;
+    static boolean verbose;
+    private static File CONFIG_FILE;
+    private static Map<String, Command> commands;
+
+    public static void init(File configFile) {
+        CONFIG_FILE = configFile;
+        config = new YamlConfiguration();
+        try {
+            config.load(CONFIG_FILE);
+        } catch (IOException ignore) {
+        } catch (InvalidConfigurationException ex) {
+            Bukkit.getLogger().log(Level.SEVERE, "Could not load zircon.yml, please correct your syntax errors", ex);
+            throw Throwables.propagate(ex);
+        }
+        config.options().header(HEADER);
+        config.options().copyDefaults(true);
+        verbose = getBoolean("verbose", false);
+
+        commands = new HashMap<>();
+        commands.put("zircon", new ZirconCommand("zircon"));
+
+        version = getInt("config-version", 1);
+        set("config-version", 1);
+
+        readConfig(ZirconConfig.class, null);
+    }
+
+    protected static void log(String s) {
+        if (verbose) {
+            log(Level.INFO, s);
+        }
+    }
+
+    protected static void log(Level level, String s) {
+        Bukkit.getLogger().log(level, s);
+    }
+
+    public static void registerCommands() {
+        for (Map.Entry<String, Command> entry : commands.entrySet()) {
+            MinecraftServer.getServer().server.getCommandMap().register(entry.getKey(), "Zircon", entry.getValue());
+        }
+    }
+
+    static void readConfig(Class<?> clazz, Object instance) {
+        for (Method method : clazz.getDeclaredMethods()) {
+            if (Modifier.isPrivate(method.getModifiers())) {
+                if (method.getParameterTypes().length == 0 && method.getReturnType() == Void.TYPE) {
+                    try {
+                        method.setAccessible(true);
+                        method.invoke(instance);
+                    } catch (InvocationTargetException ex) {
+                        throw Throwables.propagate(ex.getCause());
+                    } catch (Exception ex) {
+                        Bukkit.getLogger().log(Level.SEVERE, "Error invoking " + method, ex);
+                    }
+                }
+            }
+        }
+
+        try {
+            config.save(CONFIG_FILE);
+        } catch (IOException ex) {
+            Bukkit.getLogger().log(Level.SEVERE, "Could not save " + CONFIG_FILE, ex);
+        }
+    }
+
+    private static void set(String path, Object val) {
+        config.addDefault(path, val);
+        config.set(path, val);
+    }
+
+    private static String getString(String path, String def) {
+        config.addDefault(path, def);
+        return config.getString(path, config.getString(path));
+    }
+
+    private static boolean getBoolean(String path, boolean def) {
+        config.addDefault(path, def);
+        return config.getBoolean(path, config.getBoolean(path));
+    }
+
+    private static double getDouble(String path, double def) {
+        config.addDefault(path, def);
+        return config.getDouble(path, config.getDouble(path));
+    }
+
+    private static int getInt(String path, int def) {
+        config.addDefault(path, def);
+        return config.getInt(path, config.getInt(path));
+    }
+
+    private static <T> List getList(String path, T def) {
+        config.addDefault(path, def);
+        return config.getList(path, config.getList(path));
+    }
+
+    static Map<String, Object> getMap(String path, Map<String, Object> def) {
+        if (def != null && config.getConfigurationSection(path) == null) {
+            config.addDefault(path, def);
+            return def;
+        }
+        return toMap(config.getConfigurationSection(path));
+    }
+
+    private static Map<String, Object> toMap(ConfigurationSection section) {
+        ImmutableMap.Builder<String, Object> builder = ImmutableMap.builder();
+        if (section != null) {
+            for (String key : section.getKeys(false)) {
+                Object obj = section.get(key);
+                if (obj != null) {
+                    builder.put(key, obj instanceof ConfigurationSection val ? toMap(val) : obj);
+                }
+            }
+        }
+        return builder.build();
+    }
+}
\ No newline at end of file
diff --git a/src/main/java/me/_12emin34/zircon/ZirconWorldConfig.java b/src/main/java/me/_12emin34/zircon/ZirconWorldConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..5e74461e6d7e9ee75e94962457a54129133af36a
--- /dev/null
+++ b/src/main/java/me/_12emin34/zircon/ZirconWorldConfig.java
@@ -0,0 +1,80 @@
+package me._12emin34.zircon;
+
+import net.minecraft.server.level.ServerLevel;
+import org.apache.commons.lang.BooleanUtils;
+import org.bukkit.configuration.ConfigurationSection;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+
+import static me._12emin34.zircon.ZirconConfig.log;
+
+@SuppressWarnings("unused")
+public class ZirconWorldConfig {
+
+    private final ServerLevel level;
+    private final String worldName;
+
+    public ZirconWorldConfig(ServerLevel level, String worldName) {
+        this.level = level;
+        this.worldName = worldName;
+        init();
+    }
+
+    public void init() {
+        log("-------- World Settings For [" + worldName + "] --------");
+        ZirconConfig.readConfig(ZirconConfig.class, this);
+    }
+
+    private void set(String path, Object val) {
+        ZirconConfig.config.addDefault("world-settings.default." + path, val);
+        ZirconConfig.config.set("world-settings.default." + path, val);
+        if (ZirconConfig.config.get("world-settings." + worldName + "." + path) != null) {
+            ZirconConfig.config.addDefault("world-settings." + worldName + "." + path, val);
+            ZirconConfig.config.set("world-settings." + worldName + "." + path, val);
+        }
+    }
+
+    private ConfigurationSection getConfigurationSection(String path) {
+        ConfigurationSection section = ZirconConfig.config.getConfigurationSection("world-settings." + worldName + "." + path);
+        return section != null ? section : ZirconConfig.config.getConfigurationSection("world-settings.default." + path);
+    }
+
+    private String getString(String path, String def) {
+        ZirconConfig.config.addDefault("world-settings.default." + path, def);
+        return ZirconConfig.config.getString("world-settings." + worldName + "." + path, ZirconConfig.config.getString("world-settings.default." + path));
+    }
+
+    private boolean getBoolean(String path, boolean def) {
+        ZirconConfig.config.addDefault("world-settings.default." + path, def);
+        return ZirconConfig.config.getBoolean("world-settings." + worldName + "." + path, ZirconConfig.config.getBoolean("world-settings.default." + path));
+    }
+
+    private boolean getBoolean(String path, Predicate<Boolean> predicate) {
+        String val = getString(path, "default").toLowerCase();
+        Boolean bool = BooleanUtils.toBooleanObject(val, "true", "false", "default");
+        return predicate.test(bool);
+    }
+
+    private double getDouble(String path, double def) {
+        ZirconConfig.config.addDefault("world-settings.default." + path, def);
+        return ZirconConfig.config.getDouble("world-settings." + worldName + "." + path, ZirconConfig.config.getDouble("world-settings.default." + path));
+    }
+
+    private int getInt(String path, int def) {
+        ZirconConfig.config.addDefault("world-settings.default." + path, def);
+        return ZirconConfig.config.getInt("world-settings." + worldName + "." + path, ZirconConfig.config.getInt("world-settings.default." + path));
+    }
+
+    private <T> List<?> getList(String path, T def) {
+        ZirconConfig.config.addDefault("world-settings.default." + path, def);
+        return ZirconConfig.config.getList("world-settings." + worldName + "." + path, ZirconConfig.config.getList("world-settings.default." + path));
+    }
+
+    private Map<String, Object> getMap(String path, Map<String, Object> def) {
+        final Map<String, Object> fallback = ZirconConfig.getMap("world-settings.default." + path, def);
+        final Map<String, Object> value = ZirconConfig.getMap("world-settings." + worldName + "." + path, null);
+        return value.isEmpty() ? fallback : value;
+    }
+}
\ No newline at end of file
diff --git a/src/main/java/me/_12emin34/zircon/command/ZirconCommand.java b/src/main/java/me/_12emin34/zircon/command/ZirconCommand.java
new file mode 100644
index 0000000000000000000000000000000000000000..84f84ef31ff2b7b69d686061283eb12f54714f18
--- /dev/null
+++ b/src/main/java/me/_12emin34/zircon/command/ZirconCommand.java
@@ -0,0 +1,66 @@
+package me._12emin34.zircon.command;
+
+import me._12emin34.zircon.ZirconConfig;
+import net.minecraft.server.MinecraftServer;
+import net.minecraft.server.level.ServerLevel;
+import org.bukkit.ChatColor;
+import org.bukkit.Location;
+import org.bukkit.command.Command;
+import org.bukkit.command.CommandSender;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class ZirconCommand extends Command {
+    public ZirconCommand(String name) {
+        super(name);
+        this.description = "Zircon related commands";
+        this.usageMessage = "/zircon [reload | version]";
+        this.setPermission("bukkit.command.zircon");
+    }
+
+    @Override
+    public List<String> tabComplete(CommandSender sender, String alias, String[] args, Location location) throws IllegalArgumentException {
+        if (args.length == 1) {
+            return Stream.of("reload", "version")
+                    .filter(arg -> arg.startsWith(args[0].toLowerCase()))
+                    .collect(Collectors.toList());
+        }
+        return Collections.emptyList();
+    }
+
+    @Override
+    public boolean execute(CommandSender sender, String commandLabel, String[] args) {
+        if (!testPermission(sender)) return true;
+
+        if (args.length != 1) {
+            sender.sendMessage(ChatColor.RED + "Usage: " + usageMessage);
+            return false;
+        }
+
+        if (args[0].equalsIgnoreCase("reload")) {
+            Command.broadcastCommandMessage(sender, ChatColor.RED + "Please note that this command is not supported and may cause issues.");
+            Command.broadcastCommandMessage(sender, ChatColor.RED + "If you encounter any issues please use the /stop command to restart your server.");
+
+            MinecraftServer console = MinecraftServer.getServer();
+            ZirconConfig.init((File) console.options.valueOf("zircon-settings"));
+            for (ServerLevel level : console.getAllLevels()) {
+                level.zirconConfig.init();
+                level.resetBreedingCooldowns();
+            }
+            console.server.reloadCount++;
+
+            Command.broadcastCommandMessage(sender, ChatColor.GREEN + "Zircon config reload complete.");
+        } else if (args[0].equalsIgnoreCase("version")) {
+            Command verCmd = org.bukkit.Bukkit.getServer().getCommandMap().getCommand("version");
+            if (verCmd != null) {
+                return verCmd.execute(sender, commandLabel, new String[0]);
+            }
+        }
+
+        return true;
+    }
+}
\ No newline at end of file
diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
index d5c61053f6af91aa5d0894750ab86372a91c9a73..5502092e73a141067e1fbc7d3968045ac2c9e3dd 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedServer.java
@@ -239,6 +239,15 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         }
         net.pl3x.purpur.PurpurConfig.registerCommands();
         // Purpur end
+        // Zircon start
+        try {
+            me._12emin34.zircon.ZirconConfig.init((java.io.File) options.valueOf("zircon-settings"));
+        } catch (Exception e) {
+            DedicatedServer.LOGGER.error("Unable to load server configuration", e);
+            return false;
+        }
+        me._12emin34.zircon.ZirconConfig.registerCommands();
+        // Zircon end
         com.destroystokyo.paper.VersionHistoryManager.INSTANCE.getClass(); // load version history now
         io.papermc.paper.util.ObfHelper.INSTANCE.getClass(); // load mappings for stacktrace deobf and etc.
         io.papermc.paper.brigadier.PaperBrigadierProviderImpl.INSTANCE.getClass(); // init PaperBrigadierProvider
diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index f620ea2a4f9023490b0a57a788336f55a1fc254c..b00d5f02ca7295970af2b806a10be84b46255c97 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -170,6 +170,8 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
 
     public final net.pl3x.purpur.PurpurWorldConfig purpurConfig; // Purpur
 
+    public final me._12emin34.zircon.ZirconWorldConfig zirconConfig; // Zircon
+
     public final co.aikar.timings.WorldTimingsHandler timings; // Paper
     public static BlockPos lastPhysicsProblem; // Spigot
     private org.spigotmc.TickLimiter entityLimiter;
@@ -371,6 +373,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
         this.spigotConfig = new org.spigotmc.SpigotWorldConfig(((net.minecraft.world.level.storage.PrimaryLevelData) worlddatamutable).getLevelName()); // Spigot
         this.paperConfig = new com.destroystokyo.paper.PaperWorldConfig(((net.minecraft.world.level.storage.PrimaryLevelData) worlddatamutable).getLevelName(), this.spigotConfig); // Paper
         this.purpurConfig = new net.pl3x.purpur.PurpurWorldConfig((ServerLevel) this, ((net.minecraft.world.level.storage.PrimaryLevelData) worlddatamutable).getLevelName(), env); // Purpur
+        this.zirconConfig = new me._12emin34.zircon.ZirconWorldConfig((ServerLevel) this, ((net.minecraft.world.level.storage.PrimaryLevelData) worlddatamutable).getLevelName()); // Zircon
         this.playerBreedingCooldowns = this.getNewBreedingCooldownCache(); // Purpur
         this.generator = gen;
         this.world = new CraftWorld((ServerLevel) this, gen, biomeProvider, env);
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index e4c607e4a430f50e53ba58b92a22b7a19007b8a0..06d3e16c32ea344ef28410de99934166f7e08bda 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -925,6 +925,7 @@ public final class CraftServer implements Server {
         org.spigotmc.SpigotConfig.init((File) console.options.valueOf("spigot-settings")); // Spigot
         com.destroystokyo.paper.PaperConfig.init((File) console.options.valueOf("paper-settings")); // Paper
         net.pl3x.purpur.PurpurConfig.init((File) console.options.valueOf("purpur-settings")); // Purpur
+        me._12emin34.zircon.ZirconConfig.init((File) console.options.valueOf("zircon-settings")); // Zircon
         for (ServerLevel world : this.console.getAllLevels()) {
             // world.serverLevelData.setDifficulty(config.difficulty); // Paper - per level difficulty
             world.setSpawnSettings(world.serverLevelData.getDifficulty() != Difficulty.PEACEFUL && config.spawnMonsters, config.spawnAnimals); // Paper - per level difficulty (from MinecraftServer#setDifficulty(ServerLevel, Difficulty, boolean))
@@ -966,6 +967,7 @@ public final class CraftServer implements Server {
             world.spigotConfig.init(); // Spigot
             world.paperConfig.init(); // Paper
             world.purpurConfig.init(); // Purpur
+            world.zirconConfig.init(); // Zircon
         }
 
         Plugin[] pluginClone = pluginManager.getPlugins().clone(); // Paper
@@ -2600,6 +2602,13 @@ public final class CraftServer implements Server {
         }
         // Purpur end
 
+        // Zircon start
+        @Override
+        public YamlConfiguration getZirconConfig() {
+            return me._12emin34.zircon.ZirconConfig.config;
+        }
+        // Zircon end
+
         @Override
         public void restart() {
             org.spigotmc.RestartCommand.restart();
diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
index 56674053351962a75b2731dde137dbbc0c70dbde..533c32fa7f4a754bd2d9c9e4c99677eb07ba23ed 100644
--- a/src/main/java/org/bukkit/craftbukkit/Main.java
+++ b/src/main/java/org/bukkit/craftbukkit/Main.java
@@ -155,6 +155,14 @@ public class Main {
                     .describedAs("Yml file");
                 // Purpur end
 
+                // Zircon start
+                acceptsAll(asList("zircon", "zircon-settings"), "File for zircon settings")
+                        .withRequiredArg()
+                        .ofType(File.class)
+                        .defaultsTo(new File("zircon.yml"))
+                        .describedAs("Yml file");
+                // Purpur end
+
                 // Paper start
                 acceptsAll(asList("server-name"), "Name of the server")
                         .withRequiredArg()
